### DISCUSSION TITLE
Changing the colour of the box around tabs when rearranging, amongst others

### DISCUSSION THREAD

**Discussion by Da-Dragon** [piroor/treestyletab#3172] [Q&A] [answered]
Hello folks, I've hit a couple of road blocks when messing around with the custom style rules of TST.

<img width="319" alt="Screenshot 2022-08-22 at 08 49 38" src="https://user-images.githubusercontent.com/45169570/185857119-f8fbb10f-6a70-49bc-bd4a-b6d5abbc2d1b.png">

> Firstly: how do I change the colour of the bounding box which appears when rearranging the tabs?

<img width="213" alt="Screenshot 2022-08-22 at 08 57 31" src="https://user-images.githubusercontent.com/45169570/185858344-f73fdcf7-475f-4096-bdea-50978c180e7c.png">

> Secondly: how do I increase the spacing of the pinned items on top? I have tried changing margin-left of the pinned tab, but this seems to only be a global control that changes the spacing between the pinned elements and the border of the sidebar. Margin-right seemed to affect nothing at all.

<img width="255" alt="Screenshot 2022-08-22 at 09 03 55" src="https://user-images.githubusercontent.com/45169570/185859307-5ac6a66e-2c6a-4c59-baed-5d3b53c1f385.png">

> Lastly: The indentations seem to be slightly broken. Only the first level of indentation works properly, and implementing the fix that I found in the Q&A section did not help at all.

The Code I have: (if it helps, I based this on FlyingFox, where I removed the changes I did not want)
```#tabbar-container #tabbar{
   margin-bottom: 15px !important;
}

:root, #background{
	background:  var(--dark-base) !important;
 --tab-border-radius: 7px; /* border radius of tabs */
  --spacing: 6px; /* spacing between tabs. [<15px: compact tabs] */
   --distance-from-edge: 10px; /* distance between tabs, and left-right edges of sidebar*/
}

#all-tabs{
   margin: var(--distance-from-edge);
}

tab-item:not(.collapsed):not(.pinned){
    margin-top: var(--spacing);
    border-radius: var(--tab-border-radius);
    height: 25px;
}

.sound-button:after {
    background: var(--light-3) !important;
    margin-right: 5px;
}


tab-item:not(.active):not(.pinned){
    background: var(--dark-0) !important;
 
}

tab-item:not([data-level="0"]):not(.pinned){
   margin-left: 15px;
    margin-top: calc(var(--spacing)/2);
}

tab-item tab-favicon{
    
 	background: var(--light-1);
  	border-radius: 20%;
}


.contextual-identity-marker {
  
  width: 17px !important;
  border-radius: 20%;
   
   
}

.highlighter {
    display: none !important;
}

tab-item .label {
    color: var(--light-2) !important; /* colour of tab text */
    margin-right: 15px;
  
}

tab-item .tab-closebox {
    color: var(--light-0);
}

tab-item.collapsed{
    height: 0;
    margin: 0 !important;
    padding: 0 !important;
}

tab-item .twisty:before{
   background: var(--light-3) !important;
}

tab-item.active .label {
    color: var(--light-0) !important; /* colour of active tab text*/
}

tab-item.active {
    border-radius: 8px;
background: #2C2F32;
box-shadow: inset 9px 9px 17px var(--shadow-dark),
            inset -9px -9px 17px var(--shadow-light); !important;
}

#tabbar-container #tabbar{
   margin-bottom: 15px !important;
}

tab-item:hover {
    border-radius: 8px;
background: #2C2F32;
box-shadow: inset 9px 9px 17px var(--shadow-dark),
            inset -9px -9px 17px var(--shadow-light);
  
}

tab-item:not(pinned) tab-closebox {
   right: 0;

}



.counter{
    color: var(--light-4) !important;
    border-radius: 7px;
    margin: 2px 2px 0px 4px;
    padding: 0 4px;
    padding-bottom: 2px;
    order: -2 !important;
}

.counter:before, .counter:after{ content: ""}

/********LEFT CONTENT TAB(static)*************/

:root.left tab-item tab-favicon {    
    transform: scale(120%);
    margin-right: 5px;
}

:root.left tab-item[data-child-ids] .twisty{
    margin-right: 0px;
}


/***********NEW TAB BUTTON**********/

.newtab-button {
 	padding-top: 0px;
  border-radius: 8px;
background: var (--dark-base);
box-shadow:  7px 7px 15px var(--shadow-dark),
             -7px -7px 15px var(--shadow-light);
    margin: 0 10px!important;
  padding: 8px 0 !important;
}

.newtab-action-selector-anchor {
    border: none !important;
    margin-right: 10px;
    margin-top: 3px;
}

.newtab-button:hover {
    border-radius: 8px;
background: var (--dark-base);
box-shadow: inset 7px 7px 15px var(--shadow-dark),
            inset -7px -7px 15px var(--shadow-light);
  padding: 8px 0 !important;
}


/**********PINNED TAB***************/



tab-item.pinned {
   
    margin-left:5px;
  border-radius: 8px;
	background: var(--dark-0);
	margin-right: var(--spacing);
  
}


/***********************************/

#all-tabs {
   
}

/***********************************/

:root{ 
  --dark-base: #2C2F32; /* background */
  --dark-0: #393D41;
  --dark-1: #000000; /* non-selected tab */
  --dark-2: #FF000000;
  --dark-3: #696969; /* active tab */
  --dark-4: #242528;
  --light-0:#FFFFFF; /* selected tab text */
  --light-base: #afc0d5;
  --light-1: #DEEAF6;
  --light-2: #9E9E9E; /* tab text */
  --light-3: #87A0C0; /* dropdown */
  --light-4: #7996B9;
  --shadow-dark: #1a1c1e;
  --shadow-light:#3e4246;
  --accent: #FF0000
  --yellow: #dbc074;
  --green: #5cb8a7;
  --red: #c94f6d;
  --extension-icon-mask: grayscale(85%) invert(75%) sepia(8%) saturate(862%) hue-rotate(173deg) brightness(88%);
}
```

Thanks

**Comment by irvinm** [comment, marked as answer]
For the border of the dragged tab, it is simply a less transparent copy of the original tab so it isn't a separate property.  However, you could change the border of the tab only when you are dragging it by using the css below.  Both the tab and the moving tab will get the updated border only when dragging.  Customize to your liking.

```
tab-item.dragging tab-item-substance {
  border: solid 3px var(--active-border-color) !important;
}
```

As for the pinned tabs spacing, you should read this comment:  https://github.com/piroor/treestyletab/issues/2105#issuecomment-449974529

For the last item on the indentation, I assume that problem isn't there if you remove the CSS you added to TST?  (Is it safe to assume your CSS changes that behavior?)  It doesn't look what you pasted is complete either.  (The beginning isn't right)  In addition, I don't see any references to "tab-item-substance" which was a DOM change that was introduced in 3.8.5 and I am guessing is probably part of the issue.  (https://github.com/piroor/treestyletab/discussions/2960#discussioncomment-972745)

Oh, I think I found where you started from (Yes?  https://github.com/akshat46/FlyingFox/blob/master/treestyletab/custom.css)  I see that this was last updated Mar 21, 2021 which was BEFORE the DOM changes were done in 3.8.5 and  released on July 6, 2021.  

Finally, I do see this CSS is supposed to work with very specific set of configurations laid out here https://github.com/akshat46/FlyingFox/tree/master/treestyletab.  They even say that you need to start with the "plain" theme which is no longer part of TST initially.  (There are ways to insert that old CSS manually as well)  Maybe all of this is to say there might be a fair amount of work to get this to work as it was originally intended.  If you wanted to pull just specific sections of the CSS to work maybe we could get it to work.

**Reply by Da-Dragon** [reply]
Probably a good place as any to say that I don't have a particularly deep understanding of CSS, just what I can logically extrapolate based on a line and its effects. I used FlyingFox, specifically [this generator](https://flyingfox.netlify.app/generate), to have a list of lines that would get me close-ish to what I wanted to do, and removed pretty much everything else. Everything I pasted is what I have, if something is missing then I likely didn't know better about adding it.

1. Took the snippet as you suggested. So as I take it, I can add a box around the moving tab that I am dragging, but not around the place I am dragging it to (the black line/box indicator of whether it will be a sub-tab or placed beneath it as a level 0).

2. I took a look at the comment you linked to, with `position:fixed;top:***px;left:**px`. I am likely missing something, but it seems to stack all pinned tabs on top of one another in one position and isn't inherited by each pinned tab as a spacing between them.

3. The TST indentation itself seems to work as intended. However, the box doesn't indent with the element it surrounds, which is my issue. More than likely, I am incorrectly calling the class for which the changes should be taken. 


Likely this is even more a mess than the version I have previously: [https://pastebin.com/xN8YevFa](https://pastebin.com/xN8YevFa)

**Reply by irvinm** [reply]
So, beyond the CSS you put in pastebin, what other options are you using that FlyingFox specifies to use?  (Theme, options, etc.)  When I plug in your CSS and use the theme "no decoration" I see:

![image](https://user-images.githubusercontent.com/979729/186543201-29a0375f-9a44-4185-9bef-27ecff3852b5.png)

It might just be easier to go to "all configs" -> Export and post it so we are starting from the same spot.

**Reply by irvinm** [reply]
I cleaned it up a little and added some changes that now look ok with just this CSS and using the "no decoration" theme.

```
#tabbar-container #tabbar{
   margin-bottom: 15px !important;
}

:root, #background tab-item-substance {
	background:  var(--dark-base) !important;
 --tab-border-radius: 7px; /* border radius of tabs */
  --spacing: 6px; /* spacing between tabs. [<15px: compact tabs] */
   --distance-from-edge: 10px; /* distance between tabs, and left-right edges of sidebar*/
}

#all-tabs {
   margin: var(--distance-from-edge);
}

tab-item:not(.collapsed):not(.pinned) tab-item-substance {
    margin-top: var(--spacing);
    border-radius: var(--tab-border-radius);
    height: 25px;
}

.sound-button:after {
    background: var(--light-3) !important;
    margin-right: 5px;
}

tab-item:not(.active):not(.pinned) tab-item-substance{
    background: var(--dark-0);
}

tab-item.class1.class2.class3 tab-item-substance:not(.pinned){
   	margin-left: 5px;
    margin-top: calc(var(--spacing)/2);
}

tab-item tab-favicon{
 	background: var(--light-1);
  	border-radius: 20%;
}

.contextual-identity-marker {
  	width: 17px !important;
  	border-radius: 20%;
}

.highlighter {
    display: none !important;
}

tab-item .label {
    color: var(--light-2) !important; /* colour of tab text */
    margin-right: 15px;
}

tab-item .tab-closebox {
    color: var(--light-0);
}

tab-item.collapsed {
    height: 0;
    margin: 0 !important;
    padding: 0 !important;
}

tab-item .twisty:before {
   background: var(--light-3) !important;
}

tab-item.active .label {
    color: var(--light-0) !important; /* colour of active tab text*/
}

tab-item.active tab-item-substance {
    border-radius: 8px;
	background: #2C2F32;
	box-shadow: inset 9px 9px 17px var(--shadow-dark),
            	inset -9px -9px 17px var(--shadow-light);
}

#tabbar-container #tabbar{
   margin-bottom: 15px !important;
}

tab-item:hover tab-item-substance {
    border-radius: 8px;
	background: #2C2F32;
	box-shadow: inset 9px 9px 17px var(--shadow-dark),
           		inset -9px -9px 17px var(--shadow-light);
}

tab-item:not(pinned) tab-closebox {
   right: 0;
}

.counter {
    color: var(--light-4) !important;
    border-radius: 7px;
    margin: 2px 2px 0px 4px;
    padding: 0 4px;
    padding-bottom: 2px;
    order: -2 !important;
}

.counter:before, .counter:after{ content: ""}


/********LEFT CONTENT TAB(static)*************/

:root.left tab-item tab-favicon {    
    transform: scale(120%);
    margin-right: 5px;
}

:root.left tab-item[data-child-ids] .twisty {
    margin-right: 0px;
}


/***********NEW TAB BUTTON**********/

.newtab-button {
	padding-top: 0px;
  	border-radius: 8px;
	background: var (--dark-base);
/*
  	box-shadow:  7px 7px 15px var(--shadow-dark),
             	-7px -7px 15px var(--shadow-light);
*/
	box-shadow:  0px 0px 5px var(--shadow-dark),
             	0px 0px 5px var(--shadow-light);
    margin: 0 10px!important;
  	padding: 8px 0 !important;
}

.newtab-action-selector-anchor {
    border: none !important;
    margin-right: 10px;
    margin-top: 3px;
}

.newtab-button:hover {
    border-radius: 8px;
	background: var (--dark-base);
	box-shadow: inset 7px 7px 15px var(--shadow-dark),
    	        inset -7px -7px 15px var(--shadow-light);
  	padding: 8px 0 !important;
}


/**********PINNED TAB***************/

tab-item.pinned {
  border-radius: 8px;
	background: var(--dark-0);
	margin-right: var(--spacing);
  position:fixed;left:px
}


/***********************************/

#all-tabs {
   
}

/***********************************/

:root{ 
  --dark-base: #2C2F32; /* background */
  --dark-0: #393D41;
  --dark-1: #000000; /* non-selected tab */
  --dark-2: #FF000000;
  --dark-3: #696969; /* active tab */
  --dark-4: #242528;
  --light-0:#FFFFFF; /* selected tab text */
  --light-base: #afc0d5;
  --light-1: #DEEAF6;
  --light-2: #9E9E9E; /* tab text */
  --light-3: #87A0C0; /* dropdown */
  --light-4: #7996B9;
  --shadow-dark: #1a1c1e;
  --shadow-light:#3e4246;
  --accent: #FF0000
  --yellow: #dbc074;
  --green: #5cb8a7;
  --red: #c94f6d;
  --extension-icon-mask: grayscale(85%) invert(75%) sepia(8%) saturate(862%) hue-rotate(173deg) brightness(88%);
}
```

![image](https://user-images.githubusercontent.com/979729/186549414-6b7976fe-f23e-4de6-8b3f-456cc62e4e74.png)

**Reply by irvinm** [reply]
> 

> Took the snippet as you suggested. So as I take it, I can add a box around the moving tab that I am dragging, but not around the place I am dragging it to (the black line/box indicator of whether it will be a sub-tab or placed beneath it as a level 0).
> 
> I took a look at the comment you linked to, with position:fixed;top:***px;left:**px. I am likely missing something, but it seems to stack all pinned tabs on top of one another in one position and isn't inherited by each pinned tab as a spacing between them.
> 

As for the line\box indicator, there is an easy way to change the thickness and color of both the line and box of where you are dragging it.  Below is an example of a really thick blue target line and box.

```
:root {
  --tab-dropmarker-size: 0.5em;
  --tab-dropmarker: blue;
}
```

The other question about the snippet was meant to show that you can only change the initial margin of where to place the first pinned tab and then the rest is hard coded to lay the rest of the tabs out.  As you found out, if you try to set the margin of all of the pinned tabs to be the same, they end up on top of each other.  (I tried and saw that too)  So basically, there doesn't look to be a way to tell TST to add space between pinned tabs.

**Reply by Da-Dragon** [reply]
Thank you very much, what you posted worked out for me.

I do have one issue with the code, the part setting the background colour for the entire tab section. The code I have, with the pre DOM version of things, does what it is supposed to do, but the version from your CSS code with the new version, does seemingly nothing. Everywhere else, however, the DOM changed version fixed the issues I was having.

`:root, #background {
	background:  var(--dark-base) !important;
 --tab-border-radius: 7px; /* border radius of tabs */
  --spacing: 6px; /* spacing between tabs. [<15px: compact tabs] */
   --distance-from-edge: 10px; /* distance between tabs, and left-right edges of sidebar*/
}`
<img width="216" alt="Screenshot 2022-08-25 at 21 00 34" src="https://user-images.githubusercontent.com/45169570/186747003-99c2da2c-6374-4a68-83df-92d8f5cbba77.png">

`:root, #background tab-item-substance {
	background:  var(--dark-base) !important;
 --tab-border-radius: 7px; /* border radius of tabs */
  --spacing: 6px; /* spacing between tabs. [<15px: compact tabs] */
   --distance-from-edge: 10px; /* distance between tabs, and left-right edges of sidebar*/
}`

<img width="214" alt="Screenshot 2022-08-25 at 21 01 13" src="https://user-images.githubusercontent.com/45169570/186747129-f639b013-2a92-400d-aa88-4bdc1aaff892.png">

I'm not sure about how to export all configs, but there was nothing from Flying Fox I used outside of the TST CSS code. I do have the following changes to the userChrome.css:
```/* Hide horizontal tabs at the top of the window */
#tabbrowser-tabs {
  visibility: collapse !important;
}

#sidebar-header {
  display: none;
}

#titlebar {
  visibility: collapse;
}
```

Point 1 and 3 were fixed with the edits you made.
https://pastebin.com/JzDbLPkY Paste of current version I have

**Reply by irvinm** [reply]
You can just remove that one change I made.  Leave it the original CSS.  I guess I mistakenly thought that block was for the tab backgrounds, not for the background behind the tabs. 

```
:root, #background {
	background:  var(--dark-base) !important;
 	--tab-border-radius: 7px; /* border radius of tabs */
  	--spacing: 6px; /* spacing between tabs. [<15px: compact tabs] */
   	--distance-from-edge: 10px; /* distance between tabs, and left-right edges of sidebar*/
}
```

**Reply by Da-Dragon** [reply]
Ah, I see. Thank you again for the help you've provided me

Although I do wonder why your tab close button icon is white while mine is black. Is that due to using light mode instead of dark in the browser?

**Reply by irvinm** [reply]
Yes.  If you change Firefox to use the dark theme, the "X" will be white and the light theme will have a dark "X".

![image](https://user-images.githubusercontent.com/979729/186774024-f8a24520-71e3-4698-b0a6-04e076b11c65.png)

![image](https://user-images.githubusercontent.com/979729/186774049-697d62f8-1a33-4502-9e4c-8a56a740c444.png)

**Reply by irvinm** [reply]
@DragonGaming88 did you have any other questions?  If not, can you marked this as answered?  Thanks.

**Reply by Da-Dragon** [reply]
I think that's it.

I don't think it's possible to change the colour of an iframe unless it has those options in its configuration?
The debug menu has exactly zero information on the [extension](https://addons.mozilla.org/en-US/firefox/addon/tst-bookmarks-subpanel/), and so I don't think I can do anything about it. If I'm correct, thanks for your help with the other questions I had, and I'll mark the thread as answered.

<img width="216" alt="Screenshot 2022-08-27 at 21 37 48" src="https://user-images.githubusercontent.com/45169570/187045515-5bc347e8-9244-4ed3-995a-6c3a81f3f682.png">

**Reply by irvinm** [reply]
I wasn't able to see a way to change the iframe using TST CSS, but I think that is intentional as the iframe is intended to be independent of the parent.  In addition, there isn't a built in way (or options) to change how the subpanel is presented from the other addon.

However, you can use userContent.css to make changes but that involves more steps and isn't necessarily easy to do.

Example:

![image](https://user-images.githubusercontent.com/979729/187049101-f4c0ac88-3b34-4876-b67f-0989edf91a4a.png)
